src/masternodes/rpc_proposals.cpp

```diff
@@ -650,55 +650,180 @@ UniValue listgovproposals(const JSONRPCRequest &request) {
-        } else if (str == "voc") {
-            type = uint8_t(CPropType::VoteOfConfidence);
-        } else if (str != "all") {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "type supports cfp/voc/all");
+    if (request.params[0].isObject())
+        RPCTypeCheck(request.params, {UniValue::VOBJ}, true);
+    else
+        RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VSTR, UniValue::VNUM}, true);
+
+    uint8_t type{0}, status{0};
+    int cycle{0};
+    size_t limit         = 100;
+    CPropId start        = {};
+    bool including_start = true;
+
+    if (request.params[0].isObject() > 0) {
+        auto optionsObj = request.params[0].get_obj();
+
+        if (optionsObj.exists("type")) {
+            auto str = optionsObj["type"].get_str();
+            if (str == "cfp") {
+                type = CPropType::CommunityFundProposal;
+            } else if (str == "voc") {
+                type = CPropType::VoteOfConfidence;
+            } else if (str != "all") {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "type supports cfp/voc/all");
+            }
         }
-    }
 
-    uint8_t status{0};
-    if (request.params.size() > 1) {
-        auto str = request.params[1].get_str();
-        if (str == "voting") {
-            status = uint8_t(CPropStatusType::Voting);
-        } else if (str == "rejected") {
-            status = uint8_t(CPropStatusType::Rejected);
-        } else if (str == "completed") {
-            status = uint8_t(CPropStatusType::Completed);
-        } else if (str != "all") {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "status supports voting/rejected/completed/all");
+        if (optionsObj.exists("status")) {
+            auto str = optionsObj["status"].get_str();
+            if (str == "voting") {
+                status = CPropStatusType::Voting;
+            } else if (str == "rejected") {
+                status = CPropStatusType::Rejected;
+            } else if (str == "completed") {
+                status = CPropStatusType::Completed;
+            } else if (str != "all") {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "status supports voting/rejected/completed/all");
+            }
+        }
+
+        if (optionsObj.exists("cycle")) {
+            cycle = optionsObj["cycle"].get_int();
+            if (cycle < -1) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER,
+                                   "Incorrect cycle value (0 -> all cycles, -1 -> previous cycle, N -> nth cycle");
+            }
+        }
+
+        if (!optionsObj["limit"].isNull()) {
+            limit = (size_t)optionsObj["limit"].get_int64();
+        }
+        if (!optionsObj["start"].isNull()) {
+            including_start = false;
+            start           = ParseHashV(optionsObj["start"], "start");
+        }
+        if (!optionsObj["including_start"].isNull()) {
+            including_start = optionsObj["including_start"].getBool();
+        }
```

Comment:
What about a nested pagination to be consistent with https://github.com/DeFiCh/ain/pull/1635 ?
```suggestion
        if (!optionsObj["pagination"].isNull()) {
            UniValue paginationObj = optionsObj["pagination"].get_obj();
            if (!paginationObj["limit"].isNull()) {
                limit = (size_t)paginationObj["limit"].get_int64();
            }
            if (!paginationObj["start"].isNull()) {
                including_start  = false;
                start                   = ParseHashV(optionsObj["start"], "start");
            }
            if (!paginationObj["including_start"].isNull()) {
                including_start = paginationObj["including_start"].getBool();
            }
            if (!including_start) {
                ++start;
            }
       }
```

Reply:
Ok, I agree. I think I copy pasted from some other rpc the pagination part.